src/test/java/com/teketik/test/mockinbean/test/InterfaceImplTest.java

```diff
@@ -0,0 +1,70 @@
+package com.teketik.test.mockinbean.test;
+
+
+import com.teketik.test.mockinbean.MockInBean;
+import com.teketik.test.mockinbean.test.InterfaceImplTest.Config.LoggingService;
+import com.teketik.test.mockinbean.test.InterfaceImplTest.Config.ProviderServiceImpl;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+
+public class InterfaceImplTest extends BaseTest {
+
+    @org.springframework.boot.test.context.TestConfiguration
+    static class Config {
+
+        public interface ProviderService {
+            public String provideValue();
+        }
+
+        @Service
+        public class ProviderServiceImpl implements ProviderService {
+            @Override
+            public String provideValue() {
+                return "value1";
+            }
+        }
+
+        @Component
+        public class LoggingService {
+
+            @Autowired
+            private ProviderService providerService;
+
+            @Autowired
+            private ProviderServiceImpl providerServiceImpl;
+
+            public String logCurrentValue() {
+                return providerService.provideValue();
+            }
+
+            public String logCurrentValueWithImpl() {
+                return providerServiceImpl.provideValue();
+            }
+
+        }
+    }
+
+    @Autowired
+    protected LoggingService loggingService;
+
+    @MockInBean(LoggingService.class)
+    private ProviderServiceImpl providerService;
+
+    @MockInBean(LoggingService.class)
+    private ProviderServiceImpl providerServiceImpl;
```

Comment:
Interesting, I didn't even know having two different fields with names matching the service fields would work, I immediately assumed this would crash on multiple compatible fields found in the target service. 👍 

How about some more interesting scenarios? I think these are worth covering
- Reversed mock names to service field names - should fail on both mocks matching the same field `ProviderService providerService` in the service?
   ```java
   @MockInBean(LoggingService.class)
   ProviderService providerServiceImpl;
   @MockInBean(LoggingService.class)
   ProviderServiceImpl providerService;
   ```
- Both mock names not matching service field names - should fail on the second mock matching both fields in the service?
   ```java
   @MockInBean(LoggingService.class)
   ProviderService providerServiceMock;
   @MockInBean(LoggingService.class)
   ProviderServiceImpl providerServiceImplMock;
   ```
- Supertype mock name not matching service field name - should still work as it matches by type only, which is not assignable to the subclass?
   ```java
   @MockInBean(LoggingService.class)
   ProviderService providerServiceMock;
   @MockInBean(LoggingService.class)
   ProviderServiceImpl providerServiceImpl;
   ```
- Subtype mock name not matching service field name - should fail on the second mock matching both fields in the service?
   ```java
   @MockInBean(LoggingService.class)
   ProviderService providerService;
   @MockInBean(LoggingService.class)
   ProviderServiceImpl providerServiceImplMock;
   ```

Possibly even more once a second implementation comes into play, or even a non-direct descendant through 2 classes.

Reply:
> Interesting, I didn't even know having two different fields with names matching the service fields would work, I immediately assumed this would crash on multiple compatible fields found in the target service or having the same common supertype mocked twice. 👍
> 
> How about some more interesting scenarios? I think these are worth covering
> 
> Possibly even more once a second implementation comes into play, or even a non-direct descendant through 2 classes.

Thanks for looking at this.
There is limited test coverage on the invalid setup configurations as it happens quite early during the wiring phase.

>     * Reversed mock names to service field names - should fail on both mocks matching the same field `ProviderService providerService` in the service?
>       ```java
>       @MockInBean(LoggingService.class)
>       ProviderService providerServiceImpl;
>       @MockInBean(LoggingService.class)
>       ProviderServiceImpl providerService;
>       ```

As you mentioned, this case is not supported by mock in bean. It would fail here: https://github.com/antoinemeyer/mock-in-bean/blob/master/src/main/java/com/teketik/test/mockinbean/MockInBeanTestExecutionListener.java#L76

>     * Both mock names not matching service field names - should fail on the second mock matching both fields in the service?
>       ```java
>       @MockInBean(LoggingService.class)
>       ProviderService providerServiceMock;
>       @MockInBean(LoggingService.class)
>       ProviderServiceImpl providerServiceImplMock;
>       ```
> 

This is also not supported as the second name does not match any field and would fail here:
https://github.com/antoinemeyer/mock-in-bean/blob/master/src/main/java/com/teketik/test/mockinbean/MockInBeanTestExecutionListener.java#L62
There is no name field configuration at this point to keep usage simple (can add if there is a strong case for it).

>     * Supertype mock name not matching service field name - should still work as it matches by type only, which is not assignable to the subclass?
>       ```java
>       @MockInBean(LoggingService.class)
>       ProviderService providerServiceMock;
>       @MockInBean(LoggingService.class)
>       ProviderServiceImpl providerServiceImpl;
>       ```
> 

This does work. Added the test case.

>     * Subtype mock name not matching service field name - should fail on the second mock matching both fields in the service?
>       ```java
>       @MockInBean(LoggingService.class)
>       ProviderService providerService;
>       @MockInBean(LoggingService.class)
>       ProviderServiceImpl providerServiceImplMock;
>       ```
> 

Correct. This also fails because of mismatching name.

-----

Feel free to add more test coverage if you wish but I believe this covers already the common use cases.

Reply:
Looks better now, thanks for the fix!